File: ports/java/src/test/com/benc/BstdTest.java
package test.com.benc; // <-- CORRECT PACKAGE

// --- ADD ALL THESE IMPORTS ---
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

import java.time.Instant;
import java.util.*;
import java.util.stream.Stream;

import static main.com.benc.Bstd.*;
// Static import for all JUnit assertions
import static org.junit.jupiter.api.Assertions.*;
// --- END IMPORTS ---

@DisplayName("Bstd Library Tests (Revised)")
class BstdTest {

    @Nested
    @DisplayName("Golden Path for All Data Types")
    class DataTypesTest {

        // Helper method to reduce boilerplate and improve readability in the main test
        private int assertUnmarshalAndAdvance(int offset, byte[] buffer, Object expected, UnmarshalFunc<?> unmarshaler) throws BencException {
            UnmarshalResult<?> result = unmarshaler.unmarshal(offset, buffer);
            if (expected instanceof byte[]) {
                assertArrayEquals((byte[]) expected, (byte[]) result.value());
            } else {
                assertEquals(expected, result.value());
            }
            return result.newOffset();
        }

        @Test
        @DisplayName("Should correctly marshal, skip, and unmarshal all supported data types in sequence")
        void testAllDataTypes() throws BencException {
            // 1. Setup test data
            Random rand = new Random();
            String testStr = "Hello World!";
            byte[] testBs = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10};

            List<Object> values = Arrays.asList(
                true,
                (byte) 128,
                rand.nextFloat(),
                rand.nextDouble(),
                (long) Integer.MAX_VALUE,
                -12345L,
                (byte) -1,
                (short) -1,
                rand.nextInt(),
                rand.nextLong(),
                (long) Integer.MAX_VALUE * 2L + 1L,
                (short) 160,
                rand.nextInt() & 0xFFFFFFFFL,
                rand.nextLong(),
                testStr,
                testStr,
                testBs,
                testBs,
                Instant.ofEpochSecond(1663362895, 123456789)
            );

            // 2. Calculate total size
            int totalSize = sizeBool() + sizeByte() + sizeFloat32() + sizeFloat64() +
                sizeInt((Long) values.get(4)) + sizeInt((Long) values.get(5)) +
                sizeInt8() + sizeInt16() + sizeInt32() + sizeInt64() +
                sizeUint((Long) values.get(10)) + sizeUint16() + sizeUint32() + sizeUint64() +
                sizeString(testStr) + sizeString(testStr) +
                sizeBytes(testBs) + sizeBytes(testBs) +
                sizeTime();

            // 3. Marshal all values
            byte[] buffer = new byte[totalSize];
            int offset = 0;
            offset = marshalBool(offset, buffer, (Boolean) values.get(0));
            offset = marshalByte(offset, buffer, (Byte) values.get(1));
            offset = marshalFloat32(offset, buffer, (Float) values.get(2));
            offset = marshalFloat64(offset, buffer, (Double) values.get(3));
            offset = marshalInt(offset, buffer, (Long) values.get(4));
            offset = marshalInt(offset, buffer, (Long) values.get(5));
            offset = marshalInt8(offset, buffer, (Byte) values.get(6));
            offset = marshalInt16(offset, buffer, (Short) values.get(7));
            offset = marshalInt32(offset, buffer, (Integer) values.get(8));
            offset = marshalInt64(offset, buffer, (Long) values.get(9));
            offset = marshalUint(offset, buffer, (Long) values.get(10));
            offset = marshalUint16(offset, buffer, (Short) values.get(11));
            offset = marshalUint32(offset, buffer, ((Number) values.get(12)).intValue());
            offset = marshalUint64(offset, buffer, (Long) values.get(13));
            offset = marshalString(offset, buffer, (String) values.get(14));
            offset = marshalUnsafeString(offset, buffer, (String) values.get(15));
            offset = marshalBytes(offset, buffer, (byte[]) values.get(16));
            offset = marshalBytes(offset, buffer, (byte[]) values.get(17));
            offset = marshalTime(offset, buffer, (Instant) values.get(18));
            assertEquals(totalSize, offset, "Marshal offset should match total calculated size");

            // 4. Skip all values
            int skipOffset = 0;
            skipOffset = skipBool(skipOffset, buffer);
            skipOffset = skipByte(skipOffset, buffer);
            skipOffset = skipFloat32(skipOffset, buffer);
            skipOffset = skipFloat64(skipOffset, buffer);
            skipOffset = skipVarint(skipOffset, buffer);
            skipOffset = skipVarint(skipOffset, buffer);
            skipOffset = skipInt8(skipOffset, buffer);
            skipOffset = skipInt16(skipOffset, buffer);
            skipOffset = skipInt32(skipOffset, buffer);
            skipOffset = skipInt64(skipOffset, buffer);
            skipOffset = skipVarint(skipOffset, buffer);
            skipOffset = skipUint16(skipOffset, buffer);
            skipOffset = skipUint32(skipOffset, buffer);
            skipOffset = skipUint64(skipOffset, buffer);
            skipOffset = skipString(skipOffset, buffer);
            skipOffset = skipString(skipOffset, buffer);
            skipOffset = skipBytes(skipOffset, buffer);
            skipOffset = skipBytes(skipOffset, buffer);
            skipOffset = skipTime(skipOffset, buffer);
            assertEquals(totalSize, skipOffset, "Skip offset should match total calculated size");

            // 5. Unmarshal and verify all values
            int unmarshalOffset = 0;
            unmarshalOffset = assertUnmarshalAndAdvance(unmarshalOffset, buffer, values.get(0), Bstd::unmarshalBool);
            unmarshalOffset = assertUnmarshalAndAdvance(unmarshalOffset, buffer, values.get(1), Bstd::unmarshalByte);
            unmarshalOffset = assertUnmarshalAndAdvance(unmarshalOffset, buffer, values.get(2), Bstd::unmarshalFloat32);
            unmarshalOffset = assertUnmarshalAndAdvance(unmarshalOffset, buffer, values.get(3), Bstd::unmarshalFloat64);
            unmarshalOffset = assertUnmarshalAndAdvance(unmarshalOffset, buffer, values.get(4), Bstd::unmarshalInt);
            unmarshalOffset = assertUnmarshalAndAdvance(unmarshalOffset, buffer, values.get(5), Bstd::unmarshalInt);
            unmarshalOffset = assertUnmarshalAndAdvance(unmarshalOffset, buffer, values.get(6), Bstd::unmarshalInt8);
            unmarshalOffset = assertUnmarshalAndAdvance(unmarshalOffset, buffer, values.get(7), Bstd::unmarshalInt16);
            unmarshalOffset = assertUnmarshalAndAdvance(unmarshalOffset, buffer, values.get(8), Bstd::unmarshalInt32);
            unmarshalOffset = assertUnmarshalAndAdvance(unmarshalOffset, buffer, values.get(9), Bstd::unmarshalInt64);
            unmarshalOffset = assertUnmarshalAndAdvance(unmarshalOffset, buffer, values.get(10), Bstd::unmarshalUint);
            unmarshalOffset = assertUnmarshalAndAdvance(unmarshalOffset, buffer, values.get(11), Bstd::unmarshalUint16);
            unmarshalOffset = assertUnmarshalAndAdvance(unmarshalOffset, buffer, ((Number)values.get(12)).intValue(), Bstd::unmarshalUint32);
            unmarshalOffset = assertUnmarshalAndAdvance(unmarshalOffset, buffer, values.get(13), Bstd::unmarshalUint64);
            unmarshalOffset = assertUnmarshalAndAdvance(unmarshalOffset, buffer, values.get(14), Bstd::unmarshalString);
            unmarshalOffset = assertUnmarshalAndAdvance(unmarshalOffset, buffer, values.get(15), Bstd::unmarshalUnsafeString);
            unmarshalOffset = assertUnmarshalAndAdvance(unmarshalOffset, buffer, values.get(16), Bstd::unmarshalBytesCropped);
            unmarshalOffset = assertUnmarshalAndAdvance(unmarshalOffset, buffer, values.get(17), Bstd::unmarshalBytesCopied);
            unmarshalOffset = assertUnmarshalAndAdvance(unmarshalOffset, buffer, values.get(18), Bstd::unmarshalTime);
            assertEquals(totalSize, unmarshalOffset, "Unmarshal offset should match total buffer size");
        }
    }

    @Nested
    @DisplayName("Complex Type Tests")
    class ComplexTypesTest {

        @Test
        @DisplayName("Should correctly handle slices (Lists)")
        void testSlices() throws BencException {
            List<String> slice = Arrays.asList("slice-element-1", "slice-element-2", "slice-element-3");
            int size = sizeSlice(slice, Bstd::sizeString);
            byte[] buffer = new byte[size];

            marshalSlice(0, buffer, slice, Bstd::marshalString);

            assertEquals(size, skipSlice(0, buffer, Bstd::skipString));

            UnmarshalResult<List<String>> result = unmarshalSlice(0, buffer, Bstd::unmarshalString);
            assertEquals(size, result.newOffset());
            assertEquals(slice, result.value());
        }

        @Test
        @DisplayName("Should correctly handle maps (String keys)")
        void testMaps() throws BencException {
            Map<String, String> map = new HashMap<>();
            map.put("map-key-1", "map-value-1");
            map.put("map-key-2", "map-value-2");

            int size = sizeMap(map, Bstd::sizeString, Bstd::sizeString);
            byte[] buffer = new byte[size];

            marshalMap(0, buffer, map, Bstd::marshalString, Bstd::marshalString);

            assertEquals(size, skipMap(0, buffer, Bstd::skipString, Bstd::skipString));

            UnmarshalResult<Map<String, String>> result = unmarshalMap(0, buffer, Bstd::unmarshalString, Bstd::unmarshalString);
            assertEquals(size, result.newOffset());
            assertEquals(map, result.value());
        }
    }

    @Nested
    @DisplayName("Pointer (Nullable) Tests")
    class PointerTest {
        @Test
        @DisplayName("Should handle non-null pointers correctly")
        void testNonNilPointer() throws BencException {
            String value = "hello nullable world";
            int size = sizePointer(value, Bstd::sizeString);
            byte[] buffer = new byte[size];

            marshalPointer(0, buffer, value, Bstd::marshalString);

            assertEquals(size, skipPointer(0, buffer, Bstd::skipString));

            UnmarshalResult<String> result = unmarshalPointer(0, buffer, Bstd::unmarshalString);
            assertEquals(size, result.newOffset());
            assertEquals(value, result.value());
        }

        @Test
        @DisplayName("Should handle null pointers correctly")
        void testNilPointer() throws BencException {
            String value = null;
            int size = sizePointer(value, Bstd::sizeString);
            byte[] buffer = new byte[size];

            marshalPointer(0, buffer, value, Bstd::marshalString);

            assertEquals(size, skipPointer(0, buffer, Bstd::skipString));

            UnmarshalResult<String> result = unmarshalPointer(0, buffer, Bstd::unmarshalString);
            assertEquals(size, result.newOffset());
            assertNull(result.value());
        }
    }

    @Nested
    @DisplayName("Error Handling Tests")
    class ErrorHandlingTest {

        @FunctionalInterface
        interface UnmarshalAction { void execute(byte[] buffer) throws BencException; }

        @FunctionalInterface
        interface SkipAction { void execute(byte[] buffer) throws BencException; }

        static Stream<Arguments> bufferTooSmallProvider() {
            return Stream.of(
                Arguments.of("UnmarshalBool", (UnmarshalAction) (b) -> unmarshalBool(0, b), new byte[0]),
                Arguments.of("SkipBool", (SkipAction) (b) -> skipBool(0, b), new byte[0]),
                Arguments.of("UnmarshalFloat32", (UnmarshalAction) (b) -> unmarshalFloat32(0, b), new byte[3]),
                Arguments.of("SkipFloat32", (SkipAction) (b) -> skipFloat32(0, b), new byte[3]),
                Arguments.of("UnmarshalString", (UnmarshalAction) (b) -> unmarshalString(0, b), new byte[]{4, 'a', 'b'}),
                Arguments.of("SkipString", (SkipAction) (b) -> skipString(0, b), new byte[]{4, 'a', 'b'}),
                Arguments.of("UnmarshalSlice w/o Terminator", (UnmarshalAction) (b) -> unmarshalSlice(0, b, Bstd::unmarshalByte), new byte[]{1, 1, 1, 1}),
                Arguments.of("SkipSlice w/o Terminator", (SkipAction) (b) -> skipSlice(0, b, Bstd::skipByte), new byte[]{1, 1, 1, 1})
            );
        }

        @ParameterizedTest(name = "{0}")
        @MethodSource("bufferTooSmallProvider")
        @DisplayName("Should throw BencBufferTooSmallException")
        void testBufferTooSmall(String name, Object action, byte[] buffer) {
            if (action instanceof UnmarshalAction) {
                assertThrows(BencBufferTooSmallException.class, () -> ((UnmarshalAction) action).execute(buffer));
            } else {
                assertThrows(BencBufferTooSmallException.class, () -> ((SkipAction) action).execute(buffer));
            }
        }

        static Stream<Arguments> overflowProvider() {
            byte[] overflowBytes = new byte[10];
            Arrays.fill(overflowBytes, (byte) 0x80);
            overflowBytes[9] = 0x02;

            return Stream.of(
                Arguments.of("UnmarshalInt", (UnmarshalAction) (b) -> unmarshalInt(0, b), overflowBytes),
                Arguments.of("UnmarshalUint", (UnmarshalAction) (b) -> unmarshalUint(0, b), overflowBytes),
                Arguments.of("SkipVarint", (SkipAction) (b) -> skipVarint(0, b), overflowBytes)
            );
        }

        @ParameterizedTest(name = "{0}")
        @MethodSource("overflowProvider")
        @DisplayName("Should throw BencOverflowException for varints")
        void testVarintOverflow(String name, Object action, byte[] buffer) {
            if (action instanceof UnmarshalAction) {
                assertThrows(BencOverflowException.class, () -> ((UnmarshalAction) action).execute(buffer));
            } else {
                assertThrows(BencOverflowException.class, () -> ((SkipAction) action).execute(buffer));
            }
        }
    }
}

File: std/kotlin/src/test/kotlin/BstdTest.kt
package com.banditmoscow1337.bstd

import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import java.time.Instant
import kotlin.random.Random
import kotlin.test.assertContentEquals

@DisplayName("Bstd Library Tests")
class BstdTest {

    // --- Test: Main Data Types Roundtrip ---
    @Test
    @DisplayName("should correctly marshal, skip, and unmarshal all primary data types")
    fun testDataTypes() {
        val testStr = "Hello World!"
        val testBs = byteArrayOf(0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10)
        val negLongVal = -12345L
        val posLongVal = Long.MAX_VALUE
        // FIX: In Go, `uint` is platform-dependent but typically 64-bit on servers.
        // The correct Kotlin equivalent for testing against Go's varint encoding is ULong.
        val maxULong = ULong.MAX_VALUE

        val values = listOf(
            true,
            128.toByte(),
            Random.nextFloat(),
            Random.nextDouble(),
            posLongVal,
            negLongVal,
            (-1).toByte(),
            (-1).toShort(),
            Random.nextInt(),
            Random.nextLong(),
            maxULong,
            160.toUShort(),
            // FIX: Random.nextUInt() and nextULong() can be experimental. Use stable alternatives.
            Random.nextInt().toUInt(),
            Random.nextLong().toULong(),
            testStr,
            testBs,
        )
        
        // 1. Calculate total size
        var totalSize = 0
        for (v in values) {
            totalSize += when (v) {
                is Boolean -> Bstd.sizeBoolean()
                is Byte -> Bstd.sizeByte()
                is Float -> Bstd.sizeFloat32()
                is Double -> Bstd.sizeFloat64()
                is Long -> Bstd.sizeLongAsVarint(v) // Covers both posLongVal and negLongVal
                is Short -> Bstd.sizeInt16()
                is Int -> Bstd.sizeInt32()
                is ULong -> Bstd.sizeULongAsVarint(v) // Correctly size the ULong for varint
                is UShort -> Bstd.sizeUInt16()
                is UInt -> Bstd.sizeUInt32()
                is String -> Bstd.sizeString(v)
                is ByteArray -> Bstd.sizeBytes(v)
                else -> fail("Unhandled type in size calculation: ${v::class.simpleName}")
            }
        }

        // 2. Marshal all values into a single buffer
        val b = ByteArray(totalSize)
        var n = 0
        for (v in values) {
            n = when (v) {
                is Boolean -> Bstd.marshalBoolean(b, n, v)
                is Byte -> Bstd.marshalByte(b, n, v)
                is Float -> Bstd.marshalFloat32(b, n, v)
                is Double -> Bstd.marshalFloat64(b, n, v)
                is Long -> Bstd.marshalLongAsVarint(b, n, v)
                is Short -> Bstd.marshalInt16(b, n, v)
                is Int -> Bstd.marshalInt32(b, n, v)
                is ULong -> Bstd.marshalULongAsVarint(b, n, v)
                is UShort -> Bstd.marshalUInt16(b, n, v)
                is UInt -> Bstd.marshalUInt32(b, n, v)
                is String -> Bstd.marshalString(b, n, v)
                is ByteArray -> Bstd.marshalBytes(b, n, v)
                else -> fail("Unhandled type in marshalling: ${v::class.simpleName}")
            }
        }
        assertEquals(totalSize, n, "Marshal failed: final offset doesn't match calculated size")

        // 3. Skip all values
        var skipN = 0
        for (v in values) {
            skipN = when (v) {
                is Boolean -> Bstd.skipBoolean(b, skipN)
                is Byte -> Bstd.skipByte(b, skipN)
                is Float -> Bstd.skipFloat32(b, skipN)
                is Double -> Bstd.skipFloat64(b, skipN)
                is Long -> Bstd.skipVarint(b, skipN)
                is Short -> Bstd.skipInt16(b, skipN)
                is Int -> Bstd.skipInt32(b, skipN)
                is ULong -> Bstd.skipVarint(b, skipN)
                is UShort -> Bstd.skipUInt16(b, skipN)
                is UInt -> Bstd.skipUInt32(b, skipN)
                is String -> Bstd.skipString(b, skipN)
                is ByteArray -> Bstd.skipBytes(b, skipN)
                else -> fail("Unhandled type in skipping: ${v::class.simpleName}")
            }
        }
        assertEquals(b.size, skipN, "Skip failed: final offset doesn't match buffer length")


        // 4. Unmarshal and verify
        var unmarshalN = 0
        for (expected in values) {
            val (newN, actual) = when (expected) {
                is Boolean -> Bstd.unmarshalBoolean(b, unmarshalN)
                is Byte -> Bstd.unmarshalByte(b, unmarshalN)
                is Float -> Bstd.unmarshalFloat32(b, unmarshalN)
                is Double -> Bstd.unmarshalFloat64(b, unmarshalN)
                is Long -> Bstd.unmarshalLongAsVarint(b, unmarshalN)
                is Short -> Bstd.unmarshalInt16(b, unmarshalN)
                is Int -> Bstd.unmarshalInt32(b, unmarshalN)
                is ULong -> Bstd.unmarshalULongAsVarint(b, unmarshalN)
                is UShort -> Bstd.unmarshalUInt16(b, unmarshalN)
                is UInt -> Bstd.unmarshalUInt32(b, unmarshalN)
                is String -> Bstd.unmarshalString(b, unmarshalN)
                is ByteArray -> Bstd.unmarshalBytes(b, unmarshalN)
                else -> fail("Unhandled type in unmarshalling: ${expected::class.simpleName}")
            }

            if (expected is ByteArray) {
                assertContentEquals(expected, actual as ByteArray, "Unmarshal failed for ByteArray")
            } else {
                assertEquals(expected, actual, "Unmarshal failed for ${expected::class.simpleName}")
            }
            unmarshalN = newN
        }
        assertEquals(b.size, unmarshalN, "Unmarshal failed: final offset doesn't match buffer length")
    }

    // --- Test: Error on Buffer Too Small ---
    @Test
    @DisplayName("should throw BstdBufferTooSmallException on insufficient buffer size")
    fun testErrBufTooSmall() {
        assertThrows<BstdBufferTooSmallException> { Bstd.unmarshalBoolean(byteArrayOf(), 0) }
        assertThrows<BstdBufferTooSmallException> { Bstd.unmarshalFloat32(byteArrayOf(1,2,3), 0) }
        assertThrows<BstdBufferTooSmallException> { Bstd.unmarshalLongAsVarint(byteArrayOf(0x80.toByte()), 0) }
        assertThrows<BstdBufferTooSmallException> { Bstd.unmarshalString(byteArrayOf(2, 0), 0) }
        // FIX: The function reference must be wrapped in a lambda to match the expected signature.
        assertThrows<BstdBufferTooSmallException> { Bstd.unmarshalSlice(byteArrayOf(2,0,0,0,0), 0) { b,n -> Bstd.unmarshalByte(b,n)} }
        assertThrows<BstdBufferTooSmallException> { Bstd.skipString(byteArrayOf(2,0),0) }
    }


    // --- Test: Collections ---
    @Test
    @DisplayName("should handle slices (lists) correctly")
    fun testSlices() {
        val slice = listOf("slice-element-1", "slice-element-2", "slice-element-3")
        // FIX: Pass a lambda that matches the expected signature `(T) -> Int`
        val s = Bstd.sizeSlice(slice) { Bstd.sizeString(it) }
        val buf = ByteArray(s)
        Bstd.marshalSlice(buf, 0, slice) { b, n, t -> Bstd.marshalString(b, n, t) }

        val skippedN = Bstd.skipSlice(buf, 0) { b, n -> Bstd.skipString(b, n) }
        assertEquals(buf.size, skippedN)

        val (finalN, retSlice) = Bstd.unmarshalSlice(buf, 0) { b, n -> Bstd.unmarshalString(b, n) }
        assertEquals(buf.size, finalN)
        assertEquals(slice, retSlice)
    }

    @Test
    @DisplayName("should handle maps correctly")
    fun testMaps() {
        val m = mapOf(
            "map-key-1" to "map-value-1",
            "map-key-2" to "map-value-2"
        )
        val s = Bstd.sizeMap(m, { Bstd.sizeString(it) }, { Bstd.sizeString(it) })
        val buf = ByteArray(s)
        Bstd.marshalMap(buf, 0, m, { b, n, t -> Bstd.marshalString(b, n, t) }, { b, n, t -> Bstd.marshalString(b, n, t) })

        val skippedN = Bstd.skipMap(buf, 0, Bstd::skipString, Bstd::skipString)
        assertEquals(buf.size, skippedN)

        val (finalN, retMap) = Bstd.unmarshalMap(buf, 0, Bstd::unmarshalString, Bstd::unmarshalString)
        assertEquals(buf.size, finalN)
        assertEquals(m, retMap)
    }

    // --- Test: String Edge Cases ---
    @Test
    @DisplayName("should handle empty strings")
    fun testEmptyString() {
        val str = ""
        val s = Bstd.sizeString(str)
        val buf = ByteArray(s)
        Bstd.marshalString(buf, 0, str)

        assertEquals(1, buf.size) // Only varint '0'
        assertEquals(0.toByte(), buf[0])

        val (finalN, retStr) = Bstd.unmarshalString(buf, 0)
        assertEquals(buf.size, finalN)
        assertEquals(str, retStr)
    }

    @Test
    @DisplayName("should handle long strings")
    fun testLongString() {
        val str = "H".repeat(UShort.MAX_VALUE.toInt() + 1)
        val s = Bstd.sizeString(str)
        val buf = ByteArray(s)
        Bstd.marshalString(buf, 0, str)

        val (finalN, retStr) = Bstd.unmarshalString(buf, 0)
        assertEquals(buf.size, finalN)
        assertEquals(str, retStr)
    }

    // --- Test: Varint Edge Cases ---
    @Test
    @DisplayName("should handle varint edge cases for skipping and unmarshalling")
    fun testVarintEdgeCases() {
        val overflowBytes = ByteArray(11) { 0x80.toByte() }
        val overflowEdgeCaseBytes = ByteArray(10) { 0x80.toByte() }.apply { this[9] = 0x02 }

        // Test SkipVarint
        assertThrows<BstdOverflowException> { Bstd.skipVarint(overflowBytes, 0) }
        assertThrows<BstdOverflowException> { Bstd.skipVarint(overflowEdgeCaseBytes, 0) }
        assertThrows<BstdBufferTooSmallException> { Bstd.skipVarint(byteArrayOf(0x80.toByte()), 0) }

        // Test UnmarshalLongAsVarint and UnmarshalULongAsVarint
        assertThrows<BstdOverflowException> { Bstd.unmarshalLongAsVarint(overflowBytes, 0) }
        assertThrows<BstdOverflowException> { Bstd.unmarshalLongAsVarint(overflowEdgeCaseBytes, 0) }
        assertThrows<BstdBufferTooSmallException> { Bstd.unmarshalULongAsVarint(byteArrayOf(0x80.toByte()), 0) }
    }

    // --- Test: Time ---
    @Test
    @DisplayName("should handle Time (Instant) correctly")
    fun testTime() {
        val now = Instant.ofEpochSecond(1663362895, 123456789)
        val s = Bstd.sizeTime()
        val buf = ByteArray(s)
        Bstd.marshalTime(buf, 0, now)

        val (finalN, retTime) = Bstd.unmarshalTime(buf, 0)
        assertEquals(buf.size, finalN)
        assertEquals(now, retTime)

        // Error path
        assertThrows<BstdBufferTooSmallException> { Bstd.unmarshalTime(byteArrayOf(1, 2, 3), 0) }
    }

    // --- Test: Pointer (Nullable) ---
    @Test
    @DisplayName("should handle non-null pointer (nullable type)")
    fun testNonNilPointer() {
        val value = "hello nullable world"
        val ptr: String? = value

        val s = Bstd.sizePointer(ptr) { Bstd.sizeString(it) }
        val buf = ByteArray(s)
        Bstd.marshalPointer(buf, 0, ptr) { b, n, t -> Bstd.marshalString(b, n, t) }

        val skippedN = Bstd.skipPointer(buf, 0, Bstd::skipString)
        assertEquals(buf.size, skippedN)

        val (finalN, retPtr) = Bstd.unmarshalPointer(buf, 0, Bstd::unmarshalString)
        assertEquals(buf.size, finalN)
        assertNotNull(retPtr)
        assertEquals(value, retPtr)
    }

    @Test
    @DisplayName("should handle null pointer (nullable type)")
    fun testNilPointer() {
        val ptr: String? = null

        val s = Bstd.sizePointer(ptr) { Bstd.sizeString(it) }
        val buf = ByteArray(s)
        Bstd.marshalPointer(buf, 0, ptr) { b, n, t -> Bstd.marshalString(b, n, t) }

        assertEquals(1, buf.size)
        assertEquals(0.toByte(), buf[0])

        val (finalN, retPtr) = Bstd.unmarshalPointer(buf, 0, Bstd::unmarshalString)
        assertEquals(buf.size, finalN)
        assertNull(retPtr)
    }

    @Test
    @DisplayName("should handle errors in pointer functions")
    fun testPointerErrors() {
        // Error during bool read in skip
        assertThrows<BstdBufferTooSmallException> { Bstd.skipPointer(byteArrayOf(), 0, Bstd::skipByte) }
        // Error during bool read in unmarshal
        assertThrows<BstdBufferTooSmallException> { Bstd.unmarshalPointer(byteArrayOf(), 0, Bstd::unmarshalByte) }
        // Error during element unmarshal
        assertThrows<BstdBufferTooSmallException> { Bstd.unmarshalPointer(byteArrayOf(1), 0, Bstd::unmarshalByte) }
    }

    /**
     * **NEW TEST**
     * Verifies that the library rejects streams with incorrect terminator sequences.
     * This is a robustness improvement over the original Go library.
     */
    @Test
    @DisplayName("should throw exception for invalid terminators")
    fun testInvalidTerminator() {
        // 1. Create a valid marshalled slice
        val slice = listOf<Byte>(10, 20)
        val s = Bstd.sizeSlice(slice) { Bstd.sizeByte() }
        val buf = ByteArray(s)
        Bstd.marshalSlice(buf, 0, slice) { b, n, t -> Bstd.marshalByte(b, n, t) }

        // 2. Corrupt the terminator (e.g., change it from [1,1,1,1] to [0,0,0,0])
        buf[s - 1] = 0
        buf[s - 2] = 0

        // 3. Assert that unmarshalling now fails with the correct exception
        assertThrows<BstdInvalidTerminatorException> {
            Bstd.unmarshalSlice(buf, 0, Bstd::unmarshalByte)
        }

        // 4. Do the same for a map
        // FIX: Integer literals must be explicitly cast to Byte
        val m = mapOf<Byte, Byte>(1.toByte() to 2.toByte())
        // FIX: Pass lambdas instead of function references
        val mapSize = Bstd.sizeMap(m, { Bstd.sizeByte() }, { Bstd.sizeByte() })
        val mapBuf = ByteArray(mapSize)
        Bstd.marshalMap(mapBuf, 0, m, { b, n, t -> Bstd.marshalByte(b, n, t) }, { b, n, t -> Bstd.marshalByte(b, n, t) })
        mapBuf[mapSize - 1] = 9 // Corrupt the map terminator

        assertThrows<BstdInvalidTerminatorException> {
            Bstd.unmarshalMap(mapBuf, 0, Bstd::unmarshalByte, Bstd::unmarshalByte)
        }
    }

    // --- Test: Final Coverage items from Go test ---
    @Test
    @DisplayName("should correctly calculate size for fixed-size slices")
    fun testSizeFixedSlice() {
        val slice = listOf(10, 20, 30) // List<Int>
        val elementSize = Bstd.sizeInt32()
        val expectedSize = Bstd.sizeULongAsVarint(slice.size.toULong()) + slice.size * elementSize + 4 // 4 for terminator
        val actualSize = Bstd.sizeFixedSlice(slice, elementSize)
        assertEquals(expectedSize, actualSize)
    }
}
